src/probabilities_estimators/histograms/rectangular_binning.jl

```diff
@@ -206,19 +179,89 @@ function RectangularBinEncoding(::AbstractDataset{D, T},
+end
 
+# multiple-axis bins don't make sense for univariate input.
+function total_outcomes(::AbstractVector,symbolization::RBE{RB{Vector{Int}}})
+    msg = "total_outcomes is ambiguous for Vectors with RectangularBinning{Vector{Int}}"
+    throw(ArgumentError(msg))
+end
+total_outcomes(::AbstractVector,symbolization::RBE{RB{Int}}) =
+    symbolization.binning.ϵ
+total_outcomes(::AbstractDataset{D}, symbolization::RBE{RB{Int}}) where {D} =
+    symbolization.binning.ϵ^D
+total_outcomes(::AbstractDataset{D}, symbolization::RBE{RB{Vector{Int}}}) where {D} =
+    prod(symbolization.binning.ϵ)
 
 # When the grid is fixed by the user, we can always deduce the total number of bins,
 # even just from the binning itself - symbolization info not needed.
-const FRB = FixedRectangularBinning
-total_outcomes(b::FRB) = b.N
-total_outcomes(::AbstractVector, b::FRB) = b.N
-total_outcomes(::AbstractDataset{D, T}, b::FRB) where {D, T} = b.N^D
-total_outcomes(symbolization::RBE{B, T}) where {B <: FRB, T <: Number} =
-    symbolization.binning.N
-total_outcomes(symbolization::RBE{B, T}) where {B <: FRB, T <: SVector{D}} where D =
-    symbolization.binning.N^D
-total_outcomes(x::AbstractVector, symbolization::RBE{B, T}) where {B <: FRB, T <: Number} =
-    symbolization.binning.N
-total_outcomes(x::AbstractDataset{D}, symbolization::RBE{B, T}) where {B <: FRB, T <: SVector{D}} where D =
-    symbolization.binning.N^D
+function total_outcomes(x::Array_or_Dataset, b::FixedRectangularBinning)
+    return total_outcomes(RectangularBinEncoding(x, b))
+end
+# This function does not need `x`; all info about binning are in the encoding
+function total_outcomes(e::RectangularBinEncoding)
+    D = length(e.mini)
+    return e.binning.N^D
+end
+
+# This function does not need `x`; all info about binning are in the encoding
+function all_possible_outcomes(e::RectangularBinEncoding)
+    # We can be smart here. All possible bins are exactly the same thing
+    # as the Cartesian Indices of a NxD matrix!
+    dims = _array_dims_from_fixed_binning(e)
+    ci = CartesianIndices(dims)
+    D = length(dims)
+    outcomes = map(x -> SVector{D, Int}(Tuple(x)), vec(ci))
+    return outcomes
+end
+
+function _array_dims_from_fixed_binning(e)
+    if e.binning isa RectangularBinning
+        error("Not possible to implement for `RectangularBinning`.")
+    end
+    N = e.binning.N
+    D = length(e.mini)
+    return ntuple(i -> N, D)
+end
+
+
+##################################################################
+# low level histogram call
+##################################################################
+# This method is called by `probabilities(x::Array_or_Dataset, est::ValueHistogram)`
+"""
+    fasthist(x::Vector_or_Dataset, ϵ::AbstractBinning)
+Create an encoding for binning, then map `x` to bins, then call `fasthist!` on the bins.
+Return the output probabilities, the bins, and the created encoder.
+"""
+function fasthist(x::Vector_or_Dataset, ϵ::AbstractBinning)
```

Comment:
There's a documentation entry at the top of `utils.md` where we document "Fast histograms" using the now gone `binhist`. What is the corresponding way of obtaining this histogram now? Do we want it do be part of the public API?